ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobProvider.java

```diff
@@ -84,13 +84,24 @@ public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundEx
       return null;
     }
     ParcelFileDescriptor readSide = pipe[0];
-    ParcelFileDescriptor writeSide = pipe[1];
-
-    try (OutputStream outputStream = new ParcelFileDescriptor.AutoCloseOutputStream(writeSide)) {
-      outputStream.write(data);
-    } catch (IOException exception) {
-      return null;
-    }
+    final ParcelFileDescriptor writeSide = pipe[1];
+
+    Thread writer =
+        new Thread() {
```

Comment:
One pretty valid concern raised by @ShikaSD is that creating a new thread for each URI is a good idea can consume quite a bit of memory by accident.

What do you think about going ahead with your suggestion about conditionally creating he thread when the blob is bigger than the pipe?

Additionally, @ShikaSD suggested using an executor here to schedule things on a single thread.

Reply:
> What do you think about going ahead with your suggestion about conditionally creating he thread when the blob is bigger than the pipe?

I tried to determine the pipe capacity using `Os.fcntlInt`, but this method is only available from API 30 ([see the documentation](https://developer.android.com/reference/android/system/Os#fcntlInt(java.io.FileDescriptor,%20int,%20int))).

```java
int F_GETPIPE_SZ = 1032; // Linux 2.6.35+
FileDescriptor fd = writeSide.getFileDescriptor();
int pipeCapacity = Os.fcntlInt(fd, F_GETPIPE_SZ, 0);
```

Probably we could call `fcntl(2)` from native code instead. Another way to determine the pipe capacity would be to open and read `/proc/sys/fs/pipe-max-size`. Hopefully it's enough to assign 65536 directly for now.

I've also updated the example in RNTester so now it includes two images – one smaller and one larger than 64 KB.

<img src="https://user-images.githubusercontent.com/20516055/125651348-7f69f17f-3d1b-4b7a-845d-12054143d6a3.png" alt="Updated RNTester" width="300" />

Just to be safe, I've also checked if it works properly for an image exactly of size equal to pipe capacity. In order to generate images of arbitrary size, I've implemented a HTTP server in Flask which appends null bytes to an existing JPG image and returns the modified image in the response:

```py
import urllib

from flask import Flask, request, make_response

app = Flask(__name__)

url = 'https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/The_Earth_seen_from_Apollo_17.jpg/240px-The_Earth_seen_from_Apollo_17.jpg'
original = urllib.request.urlopen(url).read()

@app.route("/image.jpg")
def image():
    size = int(request.args['size'])
    diff = size - len(original)
    assert diff >= 0

    output = original + b'\0' * diff
    assert len(output) == size

    response = make_response(output)
    response.headers.set('Content-Type', 'image/jpeg')
    return response
```

In RNTester app I've replaced the URLs:

```jsx
<BlobImageExample
  urls={[
    'http://10.0.2.2:5000/image.jpg?size=65534',
    'http://10.0.2.2:5000/image.jpg?size=65535',
    'http://10.0.2.2:5000/image.jpg?size=65536', // max pipe capacity
    'http://10.0.2.2:5000/image.jpg?size=65537',
    'http://10.0.2.2:5000/image.jpg?size=65538',
  ]}
/>
```

<table>
<thead>
<tr>
<th>Before (always synchronous write)</th>
<th>After (conditional write)</th>
</tr>
</thead>
<tbody>
<td>
<img src="https://user-images.githubusercontent.com/20516055/125650592-3278944a-1544-4518-b5cb-e195047871a8.png" alt="Synchronous write" />
</td>
<td>
<img src="https://user-images.githubusercontent.com/20516055/125651119-22dcabe1-4aff-44de-a514-db6da91fa0b6.png" alt="Conditional write" />
</td>
</tr>
</table>

Using the new implementation with conditional write, all images are loaded properly, so it should be safe to use the condition `data.length <= PIPE_CAPACITY`.

> Additionally, @ShikaSD suggested using an executor here to schedule things on a single thread.

Good idea! Done.